[Feature] TLS authentication (#989)
ray-project/kuberay: docs/guidance/tls.md

# TLS Authentication

> Ray can be configured to use TLS on it’s gRPC channels. This means that 

Reviewer: ```suggestion
Ray can be configured to use TLS on its gRPC channels. This means that 
```

> connecting to the Ray client on the head node will require an appropriate 

Reviewer: Is this referring to [Ray Client](https://docs.ray.io/en/latest/cluster/running-applications/job-submission/ray-client.html), or is this talking about ordinary internal gRPC communication between nodes?  If it's the former, we can link to the Ray Client doc, but if not, maybe we should avoid the term "Ray Client"

Author: I just copied and pasted this paragraph from this [document](https://docs.ray.io/en/latest/ray-core/configure.html?highlight=TLS#tls-authentication).

> Ray can be configured to use TLS on it’s gRPC channels. This means that connecting to the Ray client on the head node will require an appropriate set of credentials and also that data exchanged between various processes (client, head, workers) will be encrypted.

Author: Updated https://github.com/ray-project/kuberay/pull/989/commits/0cfe9df3fe5fb17a2cd01180a06eada3a3adda51

set of credentials and also that data exchanged between various processes 
(client, head, workers) will be encrypted ([Ray's document](https://docs.ray.io/en/latest/ray-core/configure.html?highlight=tls#tls-authentication)).

> This document provides detailed instructions from generating public-private
> key pair and CA certificate to configuring KubeRay.

Reviewer: ```suggestion
This document provides detailed instructions for generating a public-private
key pair and CA certificate for configuring KubeRay.
```

* CA (certificate authority)
* CSR (certificate signing request)
* self-signed certificate

This [YouTube video](https://youtu.be/T4Df5_cojAs) is a good start.

# TL;DR

> Please note that this document is designed to support KubeRay version 0.5.0 or later. If you are using an older version of KubeRay, some of the instructions or configurations may not apply or may require additional modifications.

```sh
# Install v0.5.0 KubeRay operator
# Create a RayCluster with TLS authentication (path: kuberay/)
kubectl apply -f ray-operator/config/samples/ray-cluster.tls.yaml

# Jump to Step 4 "Verify TLS authentication" to verify the connection.
```

# Step 1 (Optional): Generate a private key and self-signed certificate for CA

In this document, a self-signed certificate is used, but users also have the
option to choose a publicly trusted certificate authority (CA) for their TLS
authentication.

```sh
# Step 1-1: Generate a self-signed certificate and a new private key file for CA.
openssl req -x509 \
            -sha256 -days 3650 \
            -nodes \
            -newkey rsa:2048 \
            -subj "/CN=*.kuberay.com/C=US/L=San Francisco" \
            -keyout ca.key -out ca.crt

> # Step 1-2: Use `cat $FILENAME | base64` to encode `ca.key` and `ca.crt`.

Author: Updated https://github.com/ray-project/kuberay/pull/989/commits/87f6afd7f095a75a3cdd5007efb7f18e9429b80c.

and the `csr.conf` configuration file.
3. A self-signed certificate (`tls.crt`) is generated using the private key of the 
Certificate Authority (`ca.key`) and the previously generated CSR.

The only difference between `gencert_head.sh` and `gencert_worker.sh` is the `[ alt_names ]`
section in `csr.conf` and `cert.conf`. The worker Pods use the fully qualified domain name 
(FQDN) of the head Kubernetes Service to establish a connection with the head Pod. 
Therefore, the `[alt_names]` section for the head Pod needs to include the FQDN of the head 
Kubernetes Service. By the way, the head Pod uses `$POD_IP` to communicate with worker Pods.

```sh
# gencert_head.sh
[alt_names]
DNS.1 = localhost
DNS.2 = $FQ_RAY_IP
IP.1 = 127.0.0.1
IP.2 = $POD_IP

# gencert_worker.sh
[alt_names]
DNS.1 = localhost
IP.1 = 127.0.0.1
IP.2 = $POD_IP
```

In [Kubernetes networking model](https://github.com/kubernetes/design-proposals-archive/blob/main/network/networking.md#pod-to-pod), the IP that a Pod sees itself as is the same IP that others see it as. That's why Ray Pods can self-register for the certificates.

# Step 3 (Optional): Configure environment variables for Ray TLS authentication

> To enable TLS authentication in your Ray cluster, set the following environment variables:

Reviewer: Can we clarify where and when these environment variables need to be set?

Author: Remove "optional" from the document, and explain what is covered by `ray-cluster.tls.yaml` instead.

https://github.com/ray-project/kuberay/pull/989/commits/d4b70a8b8e3182c014bc7cd35b33b5a320d785fb
